wasm/module.go

```diff
@@ -115,43 +103,178 @@ type Module struct {
+		ret = "f64"
 	}
+	return
+}
 
-	// Version.
-	if n, err := io.ReadFull(r, buf); err != nil || n != 4 {
-		return nil, ErrInvalidVersion
+func (t *FunctionType) String() (ret string) {
+	for _, b := range t.Params {
+		ret += valueTypeName(b)
 	}
-	for i := 0; i < 4; i++ {
-		if buf[i] != version[i] {
-			return nil, ErrInvalidVersion
-		}
+	if len(t.Params) == 0 {
+		ret += "null"
 	}
-
-	ret := &Module{}
-	if err := ret.readSections(r); err != nil {
-		return nil, fmt.Errorf("readSections failed: %w", err)
+	ret += "_"
+	for _, b := range t.Results {
+		ret += valueTypeName(b)
 	}
-
-	if len(ret.FunctionSection) != len(ret.CodeSection) {
-		return nil, fmt.Errorf("function and code section have inconsistent lengths")
+	if len(t.Results) == 0 {
+		ret += "null"
 	}
-	return ret, nil
+	return
+}
+
+// ImportKind indicates which import description is present
+// See https://www.w3.org/TR/wasm-core-1/#import-section%E2%91%A0
+type ImportKind = byte
+
+const (
+	ImportKindFunc   ImportKind = 0x00
+	ImportKindTable  ImportKind = 0x01
+	ImportKindMemory ImportKind = 0x02
+	ImportKindGlobal ImportKind = 0x03
+)
+
+// Import is the binary representation of an import indicated by Kind
+// See https://www.w3.org/TR/wasm-core-1/#binary-import
+type Import struct {
+	Kind ImportKind
+	// Module is the possibly empty primary namespace of this import
+	Module string
+	// Module is the possibly empty secondary namespace of this import
+	Name string
+	// DescFunc is the index in Module.TypeSection when Kind equals ImportKindFunc
+	DescFunc uint32
+	// DescTable is the inlined TableType when Kind equals ImportKindTable
+	DescTable *TableType
+	// DescMem is the inlined MemoryType when Kind equals ImportKindMemory
+	DescMem *MemoryType
+	// DescGlobal is the inlined GlobalType when Kind equals ImportKindGlobal
+	DescGlobal *GlobalType
+}
+
+type LimitsType struct {
+	Min uint32
+	Max *uint32
+}
+type TableType struct {
+	ElemType byte
+	Limit    *LimitsType
+}
+
+type MemoryType = LimitsType
+
+type GlobalType struct {
+	ValType ValueType
+	Mutable bool
+}
+
+type Global struct {
+	Type *GlobalType
+	Init *ConstantExpression
+}
+
+type ConstantExpression struct {
+	Opcode Opcode
+	Data   []byte
+}
+
+// ExportKind indicates which index Export.Index points to
+// See https://www.w3.org/TR/wasm-core-1/#export-section%E2%91%A0
+type ExportKind = byte
+
+const (
+	ExportKindFunc   ExportKind = 0x00
+	ExportKindTable  ExportKind = 0x01
+	ExportKindMemory ExportKind = 0x02
+	ExportKindGlobal ExportKind = 0x03
+)
+
+// Export is the binary representation of an export indicated by Kind
+// See https://www.w3.org/TR/wasm-core-1/#binary-export
+type Export struct {
+	Kind ExportKind
+	// Name is what the host refers to this definition as.
+	Name string
+	// Index is the index of the definition to export, the index namespace is by Kind
+	// Ex. If ExportKindFunc, this is an index to ModuleInstance.Functions
+	Index uint32
+}
+
+type ElementSegment struct {
```

Comment:
shouldn't we delete `segment` suffix for consistency?

Reply:
spec calls it this though, so let's leave it. I'll tidy this part up when get to the bottom of the file https://www.w3.org/TR/wasm-core-1/#element-segments%E2%91%A4